tree/dataframe/test/dataframe_datasetspec.py

```diff
@@ -0,0 +1,93 @@
+        dfWriter0.Snapshot<ULong64_t>("anotherTree", "PYspecTestFile7.root", {"z"});
+        }'''
+        ROOT.gInterpreter.Calc(code)
+
+    def tearDown(self):
+        for i in range(2, 8):
+            ROOT.gSystem.Unlink("PYspecTestFile" + str(i) + ".root")
+
+    # a test involving all 3 constructors, all possible valid ranges, all possible friends
+    def test_General(self):
+        ranges = [REntryRange(), REntryRange(1, 4), REntryRange(2,4), REntryRange(100), REntryRange(1, 100),
+                  REntryRange(2, 2), REntryRange(7, 7)]
+        expectedRess = [[100, 101, 102, 103, 104], [101, 102, 103], [102, 103], [100, 101, 102, 103, 104],
+                        [101, 102, 103, 104], [], []]
+        specs = []
+        
+        for r in ranges:
+            specs.append([])
+            specs[-1].append(RDatasetSpec("anotherTree", "PYspecTestFile7.root", r))
+            specs[-1].append(RDatasetSpec("subTree", ["PYspecTestFile2.root",
+                                                 "PYspecTestFile3.root",
+                                                 "PYspecTestFile4.root"], r))
+            specs[-1].append(RDatasetSpec([("subTree1", "PYspecTestFile5.root"),
+                                      ("subTree2", "PYspecTestFile6.root"),
+                                      ("subTree", "PYspecTestFile4.root")], r))
+
+            # for each spec add all possible combination of friends
+            for s in specs[-1]:
+                s.AddFriend("anotherTree", "PYspecTestFile7.root", "friendTree")
+                s.AddFriend("subTree", ["PYspecTestFile2.root",
+                                        "PYspecTestFile3.root",
+                                        "PYspecTestFile4.root"], "friendChain1")
+                s.AddFriend([("subTree1", "PYspecTestFile5.root"),
+                             ("subTree2", "PYspecTestFile6.root"),
+                             ("subTree", "PYspecTestFile4.root")], "friendChainN")
+
+        for i in range(len(ranges)):
+            for s in specs[i]:
+                rdf = RDataFrame(RDatasetSpec(s))
+
+                if self.legacy_pyroot:
+                    resP = rdf.Take("ULong64_t")("z")
+                    fr1P = rdf.Take("ULong64_t")("friendTree.z")
+                    fr2P = rdf.Take("ULong64_t")("friendChain1.z")
+                    fr3 = rdf.Take("ULong64_t")("friendChainN.z").GetValue()
+                    res = resP.GetValue()
+                    fr1 = fr1P.GetValue()
+                    fr2 = fr2P.GetValue()
+                else:
+                    resP = rdf.Take["ULong64_t"]("z")
+                    fr1P = rdf.Take["ULong64_t"]("friendTree.z")
+                    fr2P = rdf.Take["ULong64_t"]("friendChain1.z")
+                    fr3 = rdf.Take["ULong64_t"]("friendChainN.z").GetValue()
+                    res = resP.GetValue()
+                    fr1 = fr1P.GetValue()
+                    fr2 = fr2P.GetValue()
```

Comment:
Why is there an `if-else` if the inner statements are exactly the same?

Reply:
Take accepting either [] or () to specify the type depending if we have legacy root. But the last 4 lines can go out of the if-else.

Reply:
No, the last 4 lines cannot go out of the if else. But that's not the point, do we even need to test legacy pyroot? I would really like this to be removed.

Reply:
I don't know myself. I can just remove it ofc, and hope none of the CI build systems complains.

Reply:
If it complains we should have a way of telling the CI to avoid testing this test in case pyroot_legacy is being used. I don't know if we have any CI still using pyroot legacy

Reply:
note that this pattern is already present in `dataframe_cache.py`

Reply:
Ok then, it's outside of the scope of this PR, thanks for clarifying!

Reply:
(copied from there indeed, I am keeping it for now)